README.md
# gulp-axe-webdriver

[![Package Quality](http://npm.packagequality.com/badge/gulp-axe-webdriver.png)](http://npm.packagequality.com/badge/gulp-axe-webdriver.png)

> Gulp plugin for [aXe](https://github.com/dequelabs/axe-core) utilizing WebDriverJS.

Inspired by [grunt-axe-webdriver](https://github.com/dequelabs/grunt-axe-webdriver) and [gulp-axe-core](https://github.com/felixzapata/gulp-axe-core).

This plugin checks local and remote urls, using the Chrome browser via `chromedriver`.

## Install

```
$ npm install --save-dev gulp-axe-webdriver
```

## The task

### Usage

```js
var gulp = require('gulp');
var axe = require('gulp-axe-webdriver');

gulp.task('axe', function() {
  var options = {
    saveOutputIn: 'allHtml.json',
    urls: ['http://www.foobar-url-1/', 'http://www.foobar-url-2/']
  };
  return axe(options);
});
```

```js
var gulp = require('gulp');
var axe = require('gulp-axe-webdriver');

gulp.task('axe', function() {
  var options = {
    saveOutputIn: 'allHtml.json',
    urls: ['src/file2.html']
  };
  return axe(options);	
});
```

#### With Headless Chrome 

```js
var gulp = require('gulp');
var axe = require('gulp-axe-webdriver');

gulp.task('axe', function() {
  var options = {
    saveOutputIn: 'allHtml.json',
    headless: true,
    urls: ['src/file2.html']
  };
  return axe(options);	
});
```

#### With Glob patterns

```js
var gulp = require('gulp');
var axe = require('gulp-axe-webdriver');

gulp.task('axe', function() {
  var options = {
    saveOutputIn: 'allHtml.json',
    urls: ['src/*.html', 'http://www.foobar-url-2/']
  };
  return axe(options);
});
```

### Options
Type: `Object`

Default value:
```js
{
  errorOnViolation: false,
  folderOutputReport: 'aXeReports',
  headless: false,
  saveOutputIn: '',
  scriptTimeout: 60000,
  showOnlyViolations: false,
  tags: null,
  threshold: 0,
  urls: [],
  verbose: false
}
```

#### a11yCheckOptions
Type: `Object`

Specifies options to be used by axe.a11yCheck. Will override any other configured options. See [axe-core API documentation](https://github.com/dequelabs/axe-core/blob/master/doc/API.md) for information on its structure.

#### errorOnViolation
Type: `Boolean`

Default value: `false`

It throws an error on violation from aXe,  useful for CI environments if you want to break the build if any new violations are in your results.

#### exclude
Type: `String`

Default value: `null`

Add a CSS selector to the list of elements to exclude from analysis.

#### folderOutputReport
Type: `String`

Default value: `aXeReports`

An optional folder to indicate where the output will be saved.

#### headless
Type: `Boolean`

Default value: `false`

To run the Chrome browser in Headless mode.

#### include
Type: `String`

Default value: `null`

Adds a CSS selector to the list of elements to include in analysis.

#### saveOutputIn
Type: `String`

Default value: ''

An optional file to which the results of the accessibility scans will be written as a JSON Array of results objects.

#### scriptTimeout
Type: `Number`

Default value: `60000`

Number of milliseconds for WebDriver to wait before timing out an injected script.

#### showOnlyViolations
Type: `Boolean`

Default value: `false`

Returns only the results with the accessibility issues.

#### tags
Type: `String` or `Array[String]`

Default value: `null`

Which tags to filter violations on.

#### threshold
Type: `Number`

Default value: `0`

A number that represents the maximum number of allowable violations. Each violation represents a rule that fails, it may fail for an number of nodes. It is recommended that this value not be changed.
A negative value will prevent failure whatever the number of violations.

#### urls
Type: `Array[String]`

Default value: `[]`

An Array of URLs that will be tested. The default value is an empty array, you must supply at least one URL in order to successfully complete this task.

Can also be a glob pattern;

#### verbose
Type: `Boolean`

Default value: `false`

Show status of the analysis.

## Release History

Read the [full changelog](CHANGELOG.md).

## License

MIT © [Felix Zapata](http://github.com/felixzapata)


CHANGELOG.md
<a name="3.1.3"></a>
## [3.1.3](https://github.com/felixzapata/gulp-axe-webdriver/compare/3.1.2...3.1.3) (2019-04-24)


### Features

* **index.js:** remove import fs-path ([3ca751c](https://github.com/felixzapata/gulp-axe-webdriver/commit/3ca751c))
* **package.json:** remove fs-path package ([d2b24a5](https://github.com/felixzapata/gulp-axe-webdriver/commit/d2b24a5))



<a name="3.1.2"></a>
## [3.1.2](https://github.com/felixzapata/gulp-axe-webdriver/compare/3.1.1...3.1.2) (2019-04-20)


### Bug Fixes

* **packages:** update dependencies ([f701836](https://github.com/felixzapata/gulp-axe-webdriver/commit/f701836))



<a name="3.1.1"></a>
## [3.1.1](https://github.com/felixzapata/gulp-axe-webdriver/compare/3.1.0...3.1.1) (2019-03-15)


### Bug Fixes

* **index.js:** remove deprecated warning ([#22](https://github.com/felixzapata/gulp-axe-webdriver/issues/22)) ([23a4bcc](https://github.com/felixzapata/gulp-axe-webdriver/commit/23a4bcc))



<a name="3.1.0"></a>
# [3.1.0](https://github.com/felixzapata/gulp-axe-webdriver/compare/3.0.1...3.1.0) (2018-09-19)


### Features

* **scriptTimeout:** add scriptTimeout option ([5cc450c](https://github.com/felixzapata/gulp-axe-webdriver/commit/5cc450c)), closes [#18](https://github.com/felixzapata/gulp-axe-webdriver/issues/18)



<a name="3.0.1"></a>
## [3.0.1](https://github.com/felixzapata/gulp-axe-webdriver/compare/3.0.0...3.0.1) (2018-08-09)


### Bug Fixes

* **index.js:** return promise with axeBuilder ([20c18ed](https://github.com/felixzapata/gulp-axe-webdriver/commit/20c18ed))



<a name="3.0.0"></a>
# [3.0.0](https://github.com/felixzapata/gulp-axe-webdriver/compare/2.1.2...3.0.0) (2018-07-18)


### Features

* **index.js:** return a promise ([cb877d6](https://github.com/felixzapata/gulp-axe-webdriver/commit/cb877d6))



<a name="2.1.2"></a>
## [2.1.2](https://github.com/felixzapata/gulp-axe-webdriver/compare/2.1.1...2.1.2) (2018-06-19)

### Bug Fixes

* **index.js:** driver instance due to invalid session ID ([bbf4dd5](https://github.com/felixzapata/gulp-axe-webdriver/commit/bbf4dd5))

<a name="2.1.1"></a>
## [2.1.1](https://github.com/felixzapata/gulp-axe-webdriver/compare/2.1.0...2.1.1) (2018-05-31)


### Bug Fixes

* **index.js:** increase timeout to 60000 ms ([68d43f1](https://github.com/felixzapata/gulp-axe-webdriver/commit/68d43f1))


<a name="2.1.0"></a>
# [2.1.0](https://github.com/felixzapata/gulp-axe-webdriver/compare/2.0.0...2.1.0) (2018-04-24)


### Features

* **headless:** add support for headless mode ([7b1345b](https://github.com/felixzapata/gulp-axe-webdriver/commit/7b1345b))



<a name="2.0.0"></a>
# [2.0.0](https://github.com/felixzapata/gulp-axe-webdriver/compare/1.4.1...2.0.0) (2018-04-23)


### Features

* **package.json:** update axe-core to ^3.0.1 ([bec83e6](https://github.com/felixzapata/gulp-axe-webdriver/commit/bec83e6))
* **package.json:** update axe-webdriverjs to ^2.0.0 ([cf8e76e](https://github.com/felixzapata/gulp-axe-webdriver/commit/cf8e76e))
* **phantomjs:** remove phantomjs ([94d61e1](https://github.com/felixzapata/gulp-axe-webdriver/commit/94d61e1))



<a name="1.4.1"></a>
## [1.4.1](https://github.com/felixzapata/gulp-axe-webdriver/compare/1.4.0...1.4.1) (2018-04-23)


### Features

* **package.json:** update chromedriver to 2.38.0 ([958233b](https://github.com/felixzapata/gulp-axe-webdriver/commit/958233b))



<a name="1.4.0"></a>
# [1.4.0](https://github.com/felixzapata/gulp-axe-webdriver/compare/1.3.2...v1.4.0) (2016-12-27)


### Features

* **package.json:** update axe-webdriverjs and selenium-webdriver ([dd7aee2](https://github.com/felixzapata/gulp-axe-webdriver/commit/dd7aee2))



<a name="1.3.2"></a>
## [1.3.2](https://github.com/felixzapata/gulp-axe-webdriver/compare/1.3.1...v1.3.2) (2016-12-16)


### Bug Fixes

* **package.json:** set chromedriver to version 2.25.1 to avoid problems behind a proxy ([e99840e](https://github.com/felixzapata/gulp-axe-webdriver/commit/e99840e))



<a name="1.3.1"></a>
## [1.3.1](https://github.com/felixzapata/gulp-axe-webdriver/compare/1.3.0...v1.3.1) (2016-12-09)


### Bug Fixes

* **index.js:** modify drivers and capabilities to check complex urls in Chrome and PhantomJS ([71df07f](https://github.com/felixzapata/gulp-axe-webdriver/commit/71df07f))



<a name="1.3.0"></a>
# [1.3.0](https://github.com/felixzapata/gulp-axe-webdriver/compare/1.2.0...v1.3.0) (2016-12-09)


### Features

* **404:** create first approach to check only valid urls ([3035949](https://github.com/felixzapata/gulp-axe-webdriver/commit/3035949))
* **index.js:** filter local and non local urls to review the status ([dd2e25d](https://github.com/felixzapata/gulp-axe-webdriver/commit/dd2e25d))
* **package.json:** update axe-core to 2.0.7 ([29a3bcc](https://github.com/felixzapata/gulp-axe-webdriver/commit/29a3bcc))
* **package.json:** update axe-webdriverjs to 0.4.0 ([8eb21a2](https://github.com/felixzapata/gulp-axe-webdriver/commit/8eb21a2))
* **reporter.js:** show message when the URL is not valid ([d0be5f6](https://github.com/felixzapata/gulp-axe-webdriver/commit/d0be5f6))



<a name="1.2.0"></a>
# [1.2.0](https://github.com/felixzapata/gulp-axe-webdriver/compare/1.1.0...v1.2.0) (2016-10-10)


### Features

* **a11yCheckOptions:** add attribute a11yCheckOptions to override options ([5244f05](https://github.com/felixzapata/gulp-axe-webdriver/commit/5244f05))
* **onlyViolations:** add option to return only the violations ([14b7cc5](https://github.com/felixzapata/gulp-axe-webdriver/commit/14b7cc5))
* **verbose:** add verbose option to show information messages about the analysis ([861f076](https://github.com/felixzapata/gulp-axe-webdriver/commit/861f076))



<a name="1.1.0"></a>
# [1.1.0](https://github.com/felixzapata/gulp-axe-webdriver/compare/1.0.0...v1.1.0) (2016-09-25)


### Features

* **exclude:** add exclude option to add a CSS selector to the list of elements to exclude in analysis ([c93dd88](https://github.com/felixzapata/gulp-axe-webdriver/commit/c93dd88))

index.js
'use strict';
var path = require('path');
var fs = require('fs');
var glob = require('glob');
var AxeBuilder = require('axe-webdriverjs');
var WebDriver = require('selenium-webdriver');
var Promise = require('promise');
var fileUrl = require('file-url');
var reporter = require('./lib/reporter');
var chalk = require('chalk');
var request = require('then-request');
var PluginError = require('plugin-error');
require('chromedriver');

module.exports = function (customOptions) {

	var defaultOptions = {
		errorOnViolation: false,
		folderOutputReport: 'aXeReports',
		headless: false,
		saveOutputIn: '',
		scriptTimeout: 60000,
		showOnlyViolations: false,
		tags: null,
		threshold: 0,
		urls: [],
		verbose: false
	};

	var violationsCount = 0;
	var options = customOptions ? Object.assign(defaultOptions, customOptions) : defaultOptions;
	var chromeCapabilities = WebDriver.Capabilities.chrome();
	var chromeOptions = options.headless ? { 'args': ['--headless'] } : {};
	chromeCapabilities.set('chromeOptions', chromeOptions);
	var driver = new WebDriver.Builder().withCapabilities(chromeCapabilities).build();
	if (typeof options.scriptTimeout === 'number') {
		driver.manage().timeouts().setScriptTimeout(options.scriptTimeout);
	}
	var tagsAreDefined = (!Array.isArray(options.tags) && options.tags !== null && options.tags !== '') ||
		(Array.isArray(options.tags) && options.tags.length > 0);

	var isRemoteUrl = function (url) {
		return url.indexOf('http://') >= 0 || url.indexOf('https://') >= 0;
	};

	var flatten = function (arr) {
		return [].concat.apply([], arr);
	};

	var getUrl = function (url) {
		return isRemoteUrl(url) ? url : fileUrl(url);
	}

	var checkNotValidUrls = function (result) {
		return new Promise(function (resolve) {
			request('GET', result.url, function (error) {
				result.status = (error) ? 404 : 200;
				resolve(result);
			});
		});
	}

	var getRemoteUrls = function (result) {
		if (isRemoteUrl(result.url)) {
			return result;
		}
	};

	var getLocalUrls = function (result) {
		if (!isRemoteUrl(result.url)) {
			return result;
		}
	};

	var onlyViolations = function (item) {
		return item.violations.length > 0;
	};

	var removePassesValues = function (item) {
		delete item.passes;
		return item;
	};

	var mergeArray = function (coll, item) {
		coll.push(item);
		return coll;
	};

	var createResults = function (results) {
		var dest = '';
		var localUrls = results.filter(getLocalUrls);
		var remoteUrls = results.filter(getRemoteUrls);
		var promises = remoteUrls.map(checkNotValidUrls);
		var resultsForReporter;
		return Promise.all(promises).then(function (results) {
			resultsForReporter = localUrls.reduce(mergeArray, results);
			if (options.showOnlyViolations) {
				resultsForReporter = resultsForReporter.map(removePassesValues).filter(onlyViolations);
			}
			if (options.saveOutputIn !== '') {
				dest = path.join(options.folderOutputReport, options.saveOutputIn);
				if(!fs.existsSync(options.folderOutputReport)) {
					fs.mkdirSync(options.folderOutputReport);
				}
				fs.writeFileSync(dest, JSON.stringify(resultsForReporter, null, '  '));
			}
			if (options.verbose) {
				console.log(chalk.yellow('Preparing results'));
				console.log(chalk.yellow('================='));
			}
			reporter(resultsForReporter, options.threshold);
			if (options.errorOnViolation && violationsCount > 0) {
				throw new PluginError(
					'gulp-axe-webdriver',
					'Encountered ' + violationsCount + ' axe violation errors'
				);
			}
			return driver.quit();
		});
	};

	var findGlobPatterns = function (urls) {
		return urls.map(function (url) {
			return isRemoteUrl(url) ? url : glob.sync(url);
		});
	};

	var urls = flatten(findGlobPatterns(options.urls));

	if (options.verbose) {
		console.log(chalk.yellow('Start reading the urls'));
		console.log(chalk.yellow('======================'));
	}
	return Promise.all(urls.map(function (url) {
		return new Promise(function (resolve) {
			driver
				.get(getUrl(url))
				.then(function () {
					if (options.verbose) {
						console.log(chalk.cyan('Analysis start for: ') + url);
					}
					var startTimestamp = new Date().getTime();
					var axeBuilder = new AxeBuilder(driver);

					if (options.include) {
						axeBuilder.include(options.include);
					}

					if (options.exclude) {
						axeBuilder.exclude(options.exclude);
					}

					if (tagsAreDefined) {
						axeBuilder.withTags(options.tags);
					}

					if (options.a11yCheckOptions) {
						axeBuilder.options(options.a11yCheckOptions);
					}

					return axeBuilder.analyze(function (err, results) {
						if (err) {
							console.log(err);
						}
						results.url = url;
						results.timestamp = new Date().getTime();
						results.time = results.timestamp - startTimestamp;
						if (results.violations.length > 0) {
							++violationsCount;
						}
						if (options.verbose) {
							console.log(chalk.cyan('Analysis finished for: ') + url);
						}
						resolve(results);
					});
				});
		});

	})).then(createResults);

};
